generated/niscope/niscope/_grpc_stub_interpreter.py

```diff
@@ -0,0 +1,425 @@
+            self._client.ProbeCompensationSignalStop,
+            grpc_types.ProbeCompensationSignalStopRequest(vi=self._vi),
+        )
+
+    def read(self, channel_list, timeout, num_samples):  # noqa: N802
+        response = self._invoke(
+            self._client.Read,
+            grpc_types.ReadRequest(vi=self._vi, channel_list=channel_list, timeout=timeout, num_samples=num_samples),
+        )
+        return response.waveform, [waveform_info.WaveformInfo(x) for x in response.wfm_info]
+
+    def reset_device(self):  # noqa: N802
+        self._invoke(
+            self._client.ResetDevice,
+            grpc_types.ResetDeviceRequest(vi=self._vi),
+        )
+
+    def reset_with_defaults(self):  # noqa: N802
+        raise NotImplementedError('reset_with_defaults is not supported over gRPC')
+
+    def send_software_trigger_edge(self, which_trigger):  # noqa: N802
+        self._invoke(
+            self._client.SendSoftwareTriggerEdge,
+            grpc_types.SendSoftwareTriggerEdgeRequest(vi=self._vi, which_trigger_raw=which_trigger.value),
+        )
+
+    def set_attribute_vi_boolean(self, channel_list, attribute_id, value):  # noqa: N802
+        self._invoke(
+            self._client.SetAttributeViBoolean,
+            grpc_types.SetAttributeViBooleanRequest(vi=self._vi, channel_list=channel_list, attribute_id=attribute_id, value=value),
+        )
+
+    def set_attribute_vi_int32(self, channel_list, attribute_id, value):  # noqa: N802
+        self._invoke(
+            self._client.SetAttributeViInt32,
+            grpc_types.SetAttributeViInt32Request(vi=self._vi, channel_list=channel_list, attribute_id=attribute_id, value_raw=value),
+        )
+
+    def set_attribute_vi_int64(self, channel_list, attribute_id, value):  # noqa: N802
+        self._invoke(
+            self._client.SetAttributeViInt64,
+            grpc_types.SetAttributeViInt64Request(vi=self._vi, channel_list=channel_list, attribute_id=attribute_id, value_raw=value),
+        )
+
+    def set_attribute_vi_real64(self, channel_list, attribute_id, value):  # noqa: N802
+        self._invoke(
+            self._client.SetAttributeViReal64,
+            grpc_types.SetAttributeViReal64Request(vi=self._vi, channel_list=channel_list, attribute_id=attribute_id, value_raw=value),
+        )
+
+    def set_attribute_vi_string(self, channel_list, attribute_id, value):  # noqa: N802
+        self._invoke(
+            self._client.SetAttributeViString,
+            grpc_types.SetAttributeViStringRequest(vi=self._vi, channel_list=channel_list, attribute_id=attribute_id, value_raw=value),
+        )
+
+    def unlock(self):  # noqa: N802
+        self._lock.release()
+
+    def close(self):  # noqa: N802
+        self._invoke(
+            self._client.Close,
+            grpc_types.CloseRequest(vi=self._vi),
+        )
+
+    def error_message(self, error_code):  # noqa: N802
+        raise NotImplementedError('error_message is not supported over gRPC')
```

Comment:
See line 64 above. This method absolutely must be callable to get the message for a warning. The code is as I'd expect for the other gRPC-enabled APIs (nidmm, nifgen, niswitch).

Reply:
The method is in gRPC-device in the proto file as of the niscope 23.0.0d58 export. Has something changed?

Reply:
Ah, good catch. And this is a bit unfortunate, in grpc-device we expose and use 'GetErrorMessage' while nimi-python uses the obsolete 'ErrorMessage' method instead which we don't use in grpc-device.

They're pretty similar except `GetErrorMessage` uses ivi-dance to determine the buffer size for the message while `ErrorMessage` uses a fixed 256 size for the message.

Reply:
The reason nimi-python uses the "bad" function for messages, iirc, is that not all drivers implemented the "good" one so we had to fall back to minimum common denominator.

We could revisit this if needed. We can implement in driver runtimes that don't have them or write per-driver code when needed.

Reply:
I'm wondering here if it would be terrible for the _grpc_stub_interpreter.py for this `error_message` definition to just call grpc-device's `GetErrorMessage`. The inputs and outputs are the same, just how it calculates the error message size is different and an implementation detail. You can see nimi-python's metadata [here](https://github.com/ni/nimi-python/blob/master/src/niscope/metadata/functions.py#L3169) and grpc-device's [here for GetErrorMessage](https://github.com/ni/grpc-device/blob/main/source/codegen/metadata/niscope/functions.py#L2400)

So, I guess I could have a custom template or something that gets invoked for this NI-SCOPE to generate the call to grpc-device's `GetErrorMessage`. I guess I could also look at overriding the metadata so `grpc_name' is just 'GetErrorMessage' and 'included_in_proto' evaluates to true.

Either way, does that sound reasonable? Otherwise, I guess I'd look at exposing `GetErrorMessage` for nimi-python and using that instead of `error_message` but that seems like a more daunting change.

Reply:
Have an AzDo PR for the minimal fix up [here ](https://dev.azure.com/ni/DevCentral/_git/ni-central/pullrequest/388957)

Reply:
These fixes went in and should be pulled in by the newest update to the metadata

Reply:
I'm not thrilled with the idea of having a method that's clearly intended to be mapped to a specific C API Function calling a different C API Function. I doubt users will care, though, so I'm going to allow it.

@danestull please create an internal work item, parented under this work to follow up with @marcoskirsch when he gets back, since he never responded after his initial comment. He may have a stronger opinion and I want him to have the opportunity to voice it.

Reply:
Created item [2213312](https://dev.azure.com/ni/DevCentral/_workitems/edit/2213312) to follow up on this at a later point